### PR TITLE
[chore] Explicit IDisposable implementations for better memory management

### PR DESCRIPTION
**Pull request by nwithan8**
# Description

This PR implements the IDisposable interface to a number of in-memory objects to improve memory management and avoid orphaned objects.

The garbage collector in .NET will likely catch the majority of these objects without explicit dispose instructions, but this guarantees that no-longer-referenced objects are in fact destroyed.

Specifically:
- After every HTTP request is made, the `Request` object is no longer needed, and so is explicitly destroyed once that process is finished.
- When an `EasyPostClient` that a user would create is no longer needed (perhaps it was a one-off instance to use a specific API key), this will ensure that the `ClientConfiguration` and all `EasyPostService`s associated with that client are also destroyed.

# Testing

- N/A

# Pull Request Type

Please select the option(s) that are relevant to this PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement (fixing a typo, updating readme, renaming a variable name, etc)

### CODE REVIEW

File: EasyPost/ClientConfiguration.cs [hunk 1]
```diff
@@ -120,4 +120,37 @@ internal void SetUp()
...
+
+            // dispose of the user-provided HTTP client
+            CustomHttpClient?.Dispose();
+        }
+
+        // Free native resources (unmanaged objects) and override a finalizer below.
+        _isDisposed = true;
+    }
+
+    ~ClientConfiguration()
```
Comment (jchen293) [inline EasyPost/ClientConfiguration.cs:151]: what does `~` do here?

Comment (nwithan8) [reply EasyPost/ClientConfiguration.cs:151]: This is a deconstructor, the opposite of a constructor. It runs when an object is destroyed in memory.

File: EasyPost/ClientConfiguration.cs [hunk 2]
```diff
@@ -120,4 +120,37 @@ internal void SetUp()
 #pragma warning disable CA1307
     public override int GetHashCode() => ApiKey.GetHashCode() ^ ApiBase.GetHashCode() ^ Timeout.GetHashCode();
 #pragma warning restore CA1307
+
+    private bool _isDisposed;
+
+    public void Dispose()
```
Comment (Justintime50) [inline EasyPost/ClientConfiguration.cs:126]: suggestion: would love docstrings for these. Although "dispose" is pretty easy to follow, what exactly that does at runtime, the compiler isn't well known to everyone and could be documented.

**Review by Justintime50 — APPROVED** [review]

File: EasyPost/Client.cs [hunk 3]
```diff
@@ -94,5 +94,49 @@ public Client(ClientConfiguration configuration)
...
+        protected override void Dispose(bool disposing)
+        {
+            // ref: https://dzone.com/articles/when-and-how-to-use-dispose-and-finalize-in-c
+            // ref: https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1063#pseudo-code-example
+            if (disposing)
+            {
+                // Dispose managed state (managed objects).
+                // "disposing" inherently true when called from Dispose(), so don't need to pass it in.
+
+                // Dispose of the services
```
[the remark this answers is not in the document]
Comment (nwithan8) [reply EasyPost/Client.cs:107]: I'd rather not introduce reflection into a memory-management function.

File: EasyPost/ClientConfiguration.cs (hunk 2 above)
Comment (nwithan8) [reply EasyPost/ClientConfiguration.cs:126]: Will do in the docstrings PR.

**Pull request merged**